SHA3.cs

```diff
@@ -0,0 +1,637 @@
+    public static bool TryHashData(
+        ReadOnlySpan<byte> key,
+        ReadOnlySpan<byte> source,
+        Span<byte> destination,
+        out int bytesWritten);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(byte[] key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(ReadOnlySpan<byte> key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static int HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<int> HashDataAsync(
+        ReadOnlyMemory<byte> key,
+        Stream source,
+        Memory<byte> destination,
+        CancellationToken cancellationToken = default);
+}
+
+public abstract partial class Shake : IDisposable {
+    internal Shake(); // Will not be subclassable by developers.
+
+    public void AppendData(byte[] data);
+    public void AppendData(ReadOnlySpan<byte> data);
+    protected abstract void AppendDataCore(ReadOnlySpan<byte> data);
+
+    public byte[] GetCurrentHash(int outputLength);
+    public void GetCurrentHash(Span<byte> destination);
+    protected abstract void GetCurrentHashCore(Span<byte> destination);
+
+    public byte[] GetHashAndReset(int outputLength);
+    public void GetHashAndReset(Span<byte> destination);
+    protected abstract void GetHashAndResetCore(Span<byte> destination);
+
+    public void Dispose();
+    protected virtual void Dispose(bool disposing);
+}
+
+[UnsupportedOSPlatform("ios")]
+[UnsupportedOSPlatform("tvos")]
+[UnsupportedOSPlatform("android")]
+[UnsupportedOSPlatform("browser")]
+[UnsupportedOSPlatform("osx")]
+public sealed partial class Shake128 : Shake {
```

Comment:
Since the SHAKE APIs don't expose any statics that are useful for all OSes, and the type is sealed, we can mark the whole class as UOSP.

Reply:
Some explanation here:

1. OpenSSL gives us SHAKE128 and SHAKE256. All is good.
2. CNG gives us CSHAKE128 and CSHAKE256. But cSHAKE with empty function and customization streams is exactly the same as SHAKE. FIPS 800-185 helpfully spells this out for us:

> When N and S are both empty strings, cSHAKE(X, L, N, S) is equivalent to SHAKE as defined in FIPS 202. Thus, cSHAKE128(X, L, "", "") = SHAKE128(X, L) and cSHAKE256(X, L, "", "") = SHAKE256(X, L).

So for CNG, we will implement SHAKE in terms of CSHAKE with empty function and customization strings.